pkg/operator/operators/klusterlet/controllers/ssarcontroller/klusterlet_ssar_controller.go

```diff
@@ -143,30 +144,34 @@ func (c *ssarController) sync(ctx context.Context, controllerContext factory.Syn
 			return
 		}
 
-		bootstrapDegradedCondition := checkAgentDegradedCondition(
-			ctx, c.kubeClient,
-			operatorapiv1.ConditionHubConnectionDegraded,
-			klusterletAgent{
-				clusterName: klusterlet.Spec.ClusterName,
-				namespace:   agentNamespace,
-			},
-			klusterlet.Generation,
-			checkBootstrapSecret,
-		)
-
-		// The status is always true here since the hub kubeconfig check fails. Need to add the additional
-		// message relating to bootstrap kubeconfig check here.
-		meta.SetStatusCondition(&newKlusterlet.Status.Conditions, metav1.Condition{
-			Type:               operatorapiv1.ConditionHubConnectionDegraded,
-			Status:             metav1.ConditionTrue,
-			ObservedGeneration: klusterlet.Generation,
-			Reason:             bootstrapDegradedCondition.Reason + "," + hubConfigDegradedCondition.Reason,
-			Message:            bootstrapDegradedCondition.Message + "\n" + hubConfigDegradedCondition.Message,
-		})
-		_, err := c.patcher.PatchStatus(ctx, newKlusterlet, newKlusterlet.Status, klusterlet.Status)
-		if err != nil {
-			klog.Errorf("Update Klusterlet Status Failed: %v", err)
-			controllerContext.Queue().AddAfter(klusterletName, SSARReSyncTime)
+		// If the MultipleHubs feature is enabled, there is no need to check the bootstrap kubeconfig secret.
+		// The ssar check to bootstrapkubeconfigs is done at the start of the registration agent.
```

Comment:
I am not so sure why we need to do this. What is the impact if we alway check the bootstrap secret?

Reply:
If MultipleHubs is enabled, the legacy bootstrap-kubeconfig then is not required to be provided.

And we will get a `BootstrapSecretMissing` condition.

Reply:
can we set bootstrap-kubeconfig file as the var for this controller, so we can specify the bootstrap kubeconfig when starting the controller.

Reply:
I don't think klusterlet_ssar_controller as part of the operator can depend bootstrap-kubeconfig file. 

Using the boostrapkubeconfigs localSecrets is possible, but we should consider new condition, for example: when hub1 bootstrapkubeconfig is incorrect, but hub2 bootstrapkubeconfig is fine, do we want to consider it as a "HubConnectionDegraded" False.

Reply:
if it is not file, we can use secret? The gap is we do not know which secret is picked?

Reply:
should we just check all the secrets in this case? It will give user the info on which hub is not working right now.

Reply:
We can get the secrets from the Klusterelt.Spec, the gap is even we have the secrets, it can not act the same as the default bootstrap-hub-kubeconfig, for example:

We have a informer running against the default bootstrap-hub-kubeconfig, if the default bootstrap-hub-kubeconfig become invalid, the reconcile will trigger to update the `HubConnectionDegraded` condition: https://github.com/open-cluster-management-io/ocm/blob/a14450aa9346782a75bc4d102379715b3810740c/pkg/operator/operators/klusterlet/controllers/ssarcontroller/klusterlet_ssar_controller.go#L67 

But for the bootstrapkubeconfigs case, there is no informer for these secrets. We can add one, that means the controller will go back to watch all secrets.

---

And the `HubConnectionDegraded` condition is depended by klusterlet-controller to decide whether to deploy the work-agent: https://github.com/open-cluster-management-io/ocm/blob/a14450aa9346782a75bc4d102379715b3810740c/pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_runtime_reconcile.go#L94-L106
```go
// Deploy work agent.
	// * work agent is scaled to 0 only when degrade is true with the reason is HubKubeConfigSecretMissing.
	//   It is to ensure a fast startup of work agent when the klusterlet is bootstrapped at the first time.
	// * The work agent should not be scaled to 0 in degraded condition with other reasons,
	//   because we still need work agent running even though the hub kubconfig is missing some certain permission.
	//   It can ensure work agent to clean up the resources defined in manifestworks when cluster is detaching from the hub.
	hubConnectionDegradedCondition := meta.FindStatusCondition(klusterlet.Status.Conditions, operatorapiv1.ConditionHubConnectionDegraded)
	if hubConnectionDegradedCondition == nil {
		workConfig.Replica = 0
	} else if hubConnectionDegradedCondition.Status == metav1.ConditionTrue &&
		strings.Contains(hubConnectionDegradedCondition.Reason, operatorapiv1.ReasonHubKubeConfigSecretMissing) {
		workConfig.Replica = 0
	}
```

So, if we want to let user get the connectivity status from the condition, I suggest to add a new Condition instead of adding it to the HubConnectionDegradedCondition.

Reply:
So the questions are:
* Do we want to watch all secrets or bootstrapkubeconfigs secrets(means operator needs to add a label to the secrets configured in Klusterlet.Spec.RegistrationConfiguration.BootstrapKubeConfig.LocalSecrets)?
* Do we add want to add a new Condition to represent the connectivity status in multiplehubs case?

Reply:
> Do we want to watch all secrets or bootstrapkubeconfigs secrets(needs operator to add label on them)? 
we cannot watch all secrets. But the operator should be able to read the secrets in a certain namespace. 

> Do we want to add a new condition represent the connectivity status of bootstrapkubeconfigs?
existing status is good enough. The invalid bootstrap will not make HubConnectionDegraded to true

Reply:
> we cannot watch all secrets. But the operator should be able to read the secrets in a certain namespace.

Yes, but we get the agentNamespace from the Klusterlet by:

https://github.com/open-cluster-management-io/ocm/blob/fc9b2d96e1f49b16e03115c0d0dfa952f374dad3/pkg/operator/helpers/helpers.go#L608C1-L626C2

That means we don't know which agentnamespace the klusterlet will use at the start of the operator.

>  existing status is good enough. The invalid bootstrap will not make HubConnectionDegraded to true

Get it, just need confirm the details of the the condition, we mean:

**case1**: when multiplehubs is enabled, **hub-kubeconfig works fine**, and **none** of the bootstraps in the list is invalid, we want the HubConnectionDegraded be:

```go
metav1.Condition{
      Status: metav1.ConditionFalse,
      Reason: operatorapiv1.ReasonHubConnectionFunctional,
      Message: fmt.Sprintf("Hub kubeconfig secret %s/%s to apiserver %s is working",
			agent.namespace, helpers.HubKubeConfig, host),
}
```

**case2**: when multiplehubs is enabled, **hub-kubeconfig works fine**, and **some** of the bootstraps in the list are invalid, we want the HubConnectionDegraded be:

```go
metav1.Condition{
      Status: metav1.ConditionFalse,
      Reason: operatorapiv1.ReasonHubConnectionFunctional,
      Message: fmt.Sprintf("BootstrapKubeConfig Secrets:[%v] are getting error, %v", invaildBootstrapSecrets, err),
}
```

**case3**: when multiplehubs is enabled, **hub-kubeconfig doesn't function**, and **some** of the bootstraps in the list are invalid, we want the HubConnectionDegraded be: 

```go
metav1.Condition{
      Status: metav1.ConditionTrue,
      Reason: "BootstrapSecretsError",
      Message: fmt.Sprintf("BootstrapKubeConfig Secrets:[%v] are getting error, %v", invaildBootstrapSecrets, errs),
}
```

To me case2 is a bit odd because the reason is Functional, the status is False, but the message shows some of the BootstrapKubeConfigSecrets have trouble.

Reply:
case 2 is helpful for user to know which bootstrap kubeconfig is not working.

Reply:
That's fair, another concern is: we skip the ssar check for default bootstrapkubeconfig when hub kubeconfig is functional to avoid sending too many requests to apiserver, case2 means we are going to regression this point in the multiplehubs enabled case.

```
		// the hub kubeconfig is functional, the bootstrap kubeconfig check is not needed,
		// ignore it to avoid sending additional sar requests
		if hubConfigDegradedCondition.Status == metav1.ConditionFalse {
			meta.SetStatusCondition(&newKlusterlet.Status.Conditions, hubConfigDegradedCondition)
			_, err := c.patcher.PatchStatus(ctx, newKlusterlet, newKlusterlet.Status, klusterlet.Status)
			if err != nil {
				klog.Errorf("Update Klusterlet Status Failed: %v", err)
				controllerContext.Queue().AddAfter(klusterletName, SSARReSyncTime)
			}

			return
		}
```
https://github.com/open-cluster-management-io/ocm/blob/fc9b2d96e1f49b16e03115c0d0dfa952f374dad3/pkg/operator/operators/klusterlet/controllers/ssarcontroller/klusterlet_ssar_controller.go#L133C1-L144C4

---

If our ultimate intention is to let user know which bootstrap kubeconfig is not working, how about do it periodically? So we can 2r regression:
* Avoid watching all secrets
* Avoid sending addtional ssar when hub-kubeconfig is functional.

Reply:
To fix the first issue, how about adding a controller to add label on bootstrapkubeconfig secret, so that we can create a informer against this label?

Reply:
Decide to delay the design at the moment. 
Adding a TODO, we need to find a way to reflect the connectivity status of bootstrapkubeconfigs.